### DISCUSSION TITLE
How to manage flipped documets?

### DISCUSSION THREAD

**Discussion by dvivarelli** [deepdoctection/deepdoctection#75] [Q&A] [answered]
Hello,
First of all, this project seems ti be certo usefull for document detection task!

Now, come to the question.
I have some scanned document which could be flipped (rotated 180deg)
There Is a method to detect those and rotated it correctly?

**Comment by JaMe76** [comment, marked as answer]
Hi,

thank you for your question. 

Flipping is not supported yet, unfortunately. I haven't found a trained model, that can **classify** a page as rotated by 0,90,180,270 degrees nor an object detector that can **detect** text blocks that are rotated by 0,90,180,270 degrees. Both problems seem to be relevant though. 

Training a classifier would be not too difficult, though because you can scrape documents from the web where most pages are not rotated. When doing training one can rotate samples and update labels accordingly. One could take a pre-trained backbone from a layout detection model and then train the classifier. I think a Resnet model would be able to do the job.

Let's see if we can find someone who is willing to do the data science job and train a model. If someone provides a trained model I will be more than happy to extend the library with a pipeline component so that flipped documents can be processed to get some decent OCR results.

Apart from that I know that the OCR library DocTr can OCR results from flipped documents. I haven't tried yet...

**Reply by dvivarelli** [reply]
I have already used doctr on previous project, and from version 0.5.1 the detection phase support rotated text (using model linknet_resnet18_rotation). It returns for each bbox the rotation angle.
In the previous project, I simply estimate the rotation by median of angles.
The issue with this approch is you have to execute the detection phase first, estimate the angle, correct it and calculate the new bbox coordinate.

Otherwise i've found this project on github https://github.com/NVlabs/ocropus3-ocrorot. I'm testing it.

**Reply by JaMe76** [reply]
Sure, this will require an additional model and you will have to re-calculate bboxes. If you want to restrict to the case where pages are rotated by a multiple of 90 degrees, I think this will not require too much work within this framework because transforming coordinates happens all the time here. If you want to deal with arbitrary angles though this will be a lot more complicated because you will have an new dimension $\alpha$.  

Let us know, if the project you mention does a decent job!

**Reply by dvivarelli** [reply]
The project i have mention does not work well, I had to train a new model

**Reply by JaMe76** [reply]
Thanks for the info. 

I am implementing a pipeline component based on the solution presented in issue #30. It is rule-based and only works for small angles. Setting up a framework for solving this type of problem has to be done anyway.
